Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs

```diff
@@ -127,26 +138,61 @@ private void OnDisconnectOrTimeout(ulong clientID)
                 //We have to check here in SceneManager if our active scene is the main menu, as if it is, it means we timed out rather than a raw disconnect;
                 if (SceneManager.GetActiveScene().name != "MainMenu")
                 {
-                    // we're not at the main menu, so we obviously had a connection before... thus, we aren't in a timeout scenario.
-                    // Just shut down networking and switch back to main menu.
-                    NetworkManager.Singleton.Shutdown();
-                    if (!DisconnectReason.HasTransitionReason)
+                    if (DisconnectReason.Reason == ConnectStatus.UserRequestedDisconnect)
+                    {
+                        // simply shut down and go back to main menu
+                        NetworkManager.Singleton.Shutdown();
+                        SceneManager.LoadScene("MainMenu");
+                    }
+                    else
                     {
-                        //disconnect that happened for some other reason than user UI interaction--should display a message.
-                        DisconnectReason.SetDisconnectReason(ConnectStatus.GenericDisconnect);
+                        DisconnectReason.SetDisconnectReason(ConnectStatus.Reconnecting);
+                        // load new scene to workaround MTT-2684
+                        SceneManager.LoadSceneAsync("Loading");
+                        // try reconnecting
+                        m_TryToReconnectCoroutine ??= StartCoroutine(TryToReconnect());
                     }
-                    SceneManager.LoadScene("MainMenu");
                 }
                 else if (DisconnectReason.Reason == ConnectStatus.GenericDisconnect || DisconnectReason.Reason == ConnectStatus.Undefined)
                 {
                     // only call this if generic disconnect. Else if there's a reason, there's already code handling that popup
                     NetworkTimedOut?.Invoke();
                 }
+
                 m_ConnectStatusPub.Publish(DisconnectReason.Reason);
                 DisconnectReason.Clear();
             }
         }
 
+        private IEnumerator TryToReconnect()
+        {
+            Debug.Log("Lost connection to host, trying to reconnect...");
+            int nbTries = 0;
+            while (nbTries < 3)
+            {
+                NetworkManager.Singleton.Shutdown();
+                yield return new WaitForSeconds(1); // wait a short while to leave time for NetworkManager to complete shutting down
```

Comment:
What happens if the StopCoroutine happens during that yield? I'd do a IsConnectedClient check before the shutdown.

Reply:
Normally it shouldn't, since while connecting, we wait until k_TimeoutDuration until failing connection, so the StopCoroutine should only happen during the yield on line 177. There is maybe a small probability of the timing being too tight though, so maybe we should increase the duration of that yield to 1.1*k_timeoutDuration or something similar?